projects/core/src/lazy-loading/unified-injector.spec.ts

```diff
@@ -0,0 +1,199 @@
+    );
+  });
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('get', () => {
+    it('should emit instances from root injector', (done) => {
+      service
+        .get(TEST_TOKEN)
+        .subscribe((instance) => {
+          expect(instance).toEqual('root');
+          done();
+        })
+        .unsubscribe();
+    });
+
+    it('should emit instances if they appear in lazy loaded modules', () => {
+      const instances = [];
+      service
+        .get(TEST_TOKEN)
+        .pipe(take(3))
+        .subscribe((instance) => {
+          instances.push(instance);
+        });
+      expect(instances.length).toBe(1);
+
+      lazyModules.modules$.next(moduleInstanceWithTestToken);
+      expect(instances.length).toBe(2);
+
+      lazyModules.modules$.next(emptyModuleInstance);
+      expect(instances.length).toBe(2);
+
+      lazyModules.modules$.next(moduleInstanceWithTestToken);
+      expect(instances.length).toBe(3);
+
+      expect(instances).toEqual(['root', 'lazy', 'lazy']);
+    });
+
+    it('should emit default values if instance is missing in lazy loaded module', () => {
+      const instances = [];
+      service
+        .get(TEST_TOKEN, 'default')
+        .pipe(take(3))
+        .subscribe((instance) => {
+          instances.push(instance);
+        });
+      expect(instances.length).toBe(1);
+
+      lazyModules.modules$.next(moduleInstanceWithTestToken);
+      expect(instances.length).toBe(2);
+
+      lazyModules.modules$.next(emptyModuleInstance);
+      expect(instances.length).toBe(3);
+
+      expect(instances).toEqual(['root', 'lazy', 'default']);
+    });
+  });
+
+  describe('getMulti', () => {
+    it('should emit instances from root injector', (done) => {
+      service
+        .getMulti(TEST_MULTI_TOKEN)
+        .pipe(take(1))
+        .subscribe((instances) => {
+          expect(instances).toEqual(['root1', 'root2']);
+          done();
+        });
+    });
+
+    it('should return multi instances from root and lazy modules', (done) => {
+      lazyModules.modules$.next(moduleInstanceWithTestMultiToken);
+
+      service
+        .getMulti(TEST_MULTI_TOKEN)
+        .pipe(take(1))
+        .subscribe((instances) => {
+          expect(instances).toEqual(['root1', 'root2', 'lazy1', 'lazy2']);
+          done();
+        });
+    });
+
+    it('should re-emit new values if new instances are provided with new lazy module', (done) => {
+      service
+        .getMulti(TEST_MULTI_TOKEN)
+        .pipe(
+          delay(0), // postpone next step to next macro task to trigger asap emissions
+          tap(() => {
+            lazyModules.modules$.next(moduleInstanceWithTestMultiToken);
+          }),
+          delay(0), // postpone next step to next macro task to trigger asap emissions
+          tap(() => {
+            lazyModules.modules$.next(moduleInstanceWithTestMultiToken);
+          }),
+          take(3),
+          toArray()
```

Comment:
I have hard time undestranding the structure and correctness of this test. Especially because the `tap` operators modify the source observable. The delays also add up to the confusion.